### PR TITLE
WIP: Modularize benches

### PR DESCRIPTION
**Pull request by oxade**
1. Re-organize benchmarks
2. Use common functions for benchmarks
3. Change the way we measure latency: now we run a load gen and also send one-shot tracers to measure the latency

```
mbp> ./bench microbench latency
2022-04-12T07:44:20.665252Z  INFO sui::benchmark::transaction_creator: Open database on path: "/var/folders/3m/hnrdf4892mj0df4tqf55pm_m0000gn/T/DB_3F8D9B91E231F071D73EAEF3C197B7069B05C421"
2022-04-12T07:44:21.147115Z  INFO sui_network::transport: Listening to TCP traffic on 127.0.0.1:9555
Average Latency 3153.63 us @ 100000 tps
```

```
mbp> ./bench microbench throughput 
2022-04-12T07:44:50.747650Z  INFO sui::benchmark::transaction_creator: Open database on path: "/var/folders/3m/hnrdf4892mj0df4tqf55pm_m0000gn/T/DB_AEE3599DEC78997B6264DC322ACE9C445B7A3CA0"
2022-04-12T07:44:51.274317Z  INFO sui_network::transport: Listening to TCP traffic on 127.0.0.1:9555
Throughout: 120452.03238714246 tps
```

### CODE REVIEW

File: sui/src/bench.rs [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+
+#![deny(warnings)]
+
+use clap::*;
+
+use sui::benchmark::{bench_types, run_benchmark};
+use tracing::subscriber::set_global_default;
+use tracing_subscriber::EnvFilter;
+
+fn main() {
```
Comment (asonnino) [inline sui/src/bench.rs:12]: What are the input parameters that this function is expecting?

File: sui/src/benchmark.rs [hunk 2]
```diff
@@ -0,0 +1,289 @@
...
+}
+
+fn run_microbenchmark(benchmark: Benchmark) -> MicroBenchmarkResult {
+    #[allow(irrefutable_let_patterns)]
+    let (host, port, type_) =
+        if let BenchmarkType::MicroBenchmark { host, port, type_ } = benchmark.bench_type {
+            (host, port, type_)
+        } else {
+            panic!("Invalid variant")
+        };
```
[the remark this answers is not in the document]
Comment (asonnino) [reply sui/src/benchmark.rs:41]: can this work?

File: sui/src/benchmark.rs [hunk 3]
```diff
@@ -0,0 +1,289 @@
...
+        0,
+        "num_transactions must integer divide batch_size",
+    );
+
+    assert!(
+        (num_transactions % connections) == 0,
+        "num_transactions must {} be multiple of number of TCP connections {}",
+        num_transactions,
+        connections
+    );
```
Comment (asonnino) [inline sui/src/benchmark.rs:107]: Should we add comments to say why we need those checks?

File: sui/src/benchmark.rs [hunk 4]
```diff
@@ -0,0 +1,289 @@
...
+            let server = spawn_authority_server(network_server, tx_cr.authority_state).await;
+            if let Err(e) = server.join().await {
+                error!("Server ended with an error: {e}");
+            }
+        });
+    });
+
+    let runtime = Builder::new_multi_thread()
+        .enable_all()
+        .thread_stack_size(32 * 1024 * 1024)
```
Comment (asonnino) [inline sui/src/benchmark.rs:136]: Are these params always the same as the ones in line 122? If so should they be in common?

File: sui/src/benchmark.rs [hunk 5]
```diff
@@ -0,0 +1,289 @@
...
+    let load_gen_txes = tx_cr.generate_transactions(connections, use_move, chunk_size, num_chunks);
+
+    // These are tracer TXes used for measuring latency
+    let tracer_txes = tx_cr.generate_transactions(1, use_move, 1, num_chunks);
+
+    // Make multi-threaded runtime for the authority
+    thread::spawn(move || {
+        let runtime = Builder::new_multi_thread()
+            .enable_all()
+            .thread_stack_size(32 * 1024 * 1024)
```
Comment (asonnino) [inline sui/src/benchmark.rs:201]: same comment

File: sui/src/benchmark.rs [hunk 6]
```diff
@@ -0,0 +1,289 @@
...
+    let _batch_client_handle = tokio::task::spawn(async move {
+        let authority_client = AuthorityClient::new(network_client);
+
+        let mut start = 0;
+
+        loop {
+            let receiver = authority_client
+                .handle_batch_streaming_as_stream(BatchInfoRequest {
+                    start,
+                    end: start + 10_000,
```
Comment (asonnino) [inline sui/src/benchmark.rs:259]: What is the 10_000?

File: sui/src/benchmark/bench_types.rs [hunk 7]
```diff
@@ -0,0 +1,153 @@
...
+                // Average the latency. Probably not the best idea since they vary
+                // Should probably do better stats on these numbers
+                let tracer_avg =
+                    chunk_latencies.iter().sum::<u128>() as f64 / chunk_latencies.len() as f64;
+
+                write!(
+                    f,
+                    "Average Latency {} us @ {} tps",
+                    tracer_avg,
+                    1_000_000.0 * *load_chunk_size as f64 / *tick_period as f64,
```
Comment (asonnino) [inline sui/src/benchmark/bench_types.rs:148]: This seems the same computation as before in benchmark.rs

File: sui/src/benchmark.rs [hunk 8]
```diff
@@ -0,0 +1,289 @@
...
+        });
+    });
+
+    // Wait for server start
+    sleep(Duration::from_secs(3));
+
+    let runtime = Builder::new_multi_thread()
+        .enable_all()
+        .thread_stack_size(32 * 1024 * 1024)
+        .worker_threads(usize::min(num_cpus::get(), 24))
```
Comment (mystenmark) [inline sui/src/benchmark.rs:219]: Depending on where this will be run, we might get more reproducible results if we use, say, half the CPUs.

Comment (mystenmark) [reply sui/src/benchmark.rs:219]: (Also the authority should probably use only half as well, looks like its using all cores right now)

Comment (oxade) [reply sui/src/benchmark.rs:219]: I tried with half and still got same level of variance

File: sui/src/bench.rs (hunk 1 above)
Comment (oxade) [reply sui/src/bench.rs:12]: It uses CLI args. Am I supposed to specify something here?

File: sui/src/benchmark.rs (hunk 6 above)
Comment (oxade) [reply sui/src/benchmark.rs]: This was added by @gdanezis 
I simply ported it over.
In reality we want to benchmark this follower logic more thoroughly but thats still in the works

File: sui/src/bench.rs (hunk 1 above)
Comment (oxade) [reply sui/src/bench.rs:12]: https://github.com/MystenLabs/sui/blob/b14b4a7c4c74675ca539ddd7fb27f2ba00007af8/sui/src/benchmark/bench_types.rs

Comment (oxade) [reply sui/src/bench.rs:12]: @asonnino https://github.com/MystenLabs/sui/blob/b14b4a7c4c74675ca539ddd7fb27f2ba00007af8/sui/src/benchmark/bench_types.rs

**Review by asonnino — APPROVED** [review]

**Pull request merged**